File: Semester 4/Code Optimization and Debugging-II/Python/Exp_11 - Insertion Sort.py
# Python Program for Insertion Sort

# Eg:
# n = 7 
# Array Elements = {70, 50, 30, 10, 60, 40, 20}
# Sorted Array = {10, 20, 30, 40, 50, 60, 70}

def insertionSort(T1):
    for i in range(1, len(T1)):
        temp = T1[i]
        j = i - 1
        while (j >= 0 and temp < T1[j]):
            T1[j + 1] = T1[j]
            j = j - 1
        T1[j + 1] = temp


# n = int(input()) temporarily
n = 6

T1 = input().split()
T2 = T1
T1 = [int(x) for x in T1]
insertionSort(T1)
for i in range(len(T1)):
   print (T1[i],end=' ')

# (1)
# Input: 7
#        70 50 30 10 60 40 20
# Output: 10 20 30 40 50 60 70
#
# (2)
# Input: 6
#        66 44 22 11 33 55
# Output: 11 22 33 44 55 66

File: Semester 4/Code Optimization and Debugging-II/C++/Exp_11 - Insertion Sort.cpp
// C++ Program for Insertion Sort

/* Eg:
n = 7
Array Elements = {70, 50, 30, 10, 60, 20, 40}
Sorted Array = {10, 20, 30, 40, 50, 60, 70}
*/

#include <iostream>
using std::cout;
using std::cin;

void insertionSort(int *array, int size) {
   int k, j;
   for(int i = 1; i < size; i++) {
      k = array[i];
      j = i;
      while(j > 0 && array[j - 1] > k) {
         array[j] = array[j - 1];
         j--;
      }
      array[j] = k;
   }
}

void display(int *array, int size) {
   for(int i = 0; i < size; i++)
      cout << array[i] << " ";
   cout << "\n";
}

int main() {
   int n = 6;
   // cin >> n; temporarily
   int arr[n];
   for(int i = 0; i < n; i++) {
      cin >> arr[i];
   }
   insertionSort(arr, n);
   display(arr, n);
}

/* (1)
   Input:7
         70 50 30 10 60 20 40
   Output: 10 20 30 40 50 60 70
   
   (2)
   Input: 6
          66 44 22 11 33 55
   Output: 11 22 33 44 55 66
*/
